### PR TITLE
Switch back to protos over HTTP, instead of grpc.

### LINKED ISSUE #1982: RFC: Consider alternatives to GRPC for generic write path?
My aim is to support the new grpc generic write path in Frankenstein.  On the surface this seems easy - however I've hit a number of problems that make me think it might be better to not use grpc just yet.

The explanation of the problems requires a little background.  At weave, traffic to frankenstein need to go through a couple of services first, for SSL and to be authenticated.  So traffic goes:

    internet -> frontend -> authfe -> frankenstein

- The frontend is Nginx, and adds/removes SSL.  Its done this way for legacy reasons, so the certs can be managed in one place, although eventually we imagine we'll merge it with authfe.  All traffic from frontend is sent to authfe.
- Authfe checks the auth tokens / cookie etc and then picks the service to forward the RPC to.
- Frankenstein accepts the reads and does the right thing with them.

First problem I hit was Nginx won't proxy http2 requests - it can accept them, but all calls downstream are http1 (see https://trac.nginx.org/nginx/ticket/923).  This wasn't such a big deal, so it now looks like:

    internet --(grpc/http2)--> frontend --(grpc/http1)--> authfe --(grpc/http1)--> frankenstein

Next problem was golang grpc server won't accept http1 requests (see https://groups.google.com/forum/#!topic/grpc-io/JnjCYGPMUms).  It is possible to link a grpc server in with a normal go http mux, as long as the mux server is serving over SSL, as the golang http client & server won't do http2 over anything other than an SSL connection.  This would require making all our service to service comms SSL.  So I had a go a writing a grpc http1 server, and got pretty far.  But is was a bit of a mess.

So finally I thought I'd make a separate grpc frontend for this, running in parallel with the frontend/authfe combo on a different port - and first up I'd need a grpc reverse proxy.  Ideally we'd have some nice, generic reverse proxy that only knew about a map from service names -> downstream service, and didn't need to decode & re-encode every request as it went through.  It seems like this can't be done with golang's grpc library - see https://github.com/mwitkow/grpc-proxy/issues/1.

And then I was surprised to find you can't do grpc from browsers! See http://www.grpc.io/faq/ - not important to us, but I'm starting to question why we decided to use grpc in the first place?

It would seem we could have most of the benefits of grpc with protos over HTTP, and this wouldn't preclude moving to grpc when its a bit more mature?  In fact, the grcp FAQ even admits as much:

> Why is gRPC better than any binary blob over HTTP/2?
> This is largely what gRPC is on the wire.

### PR DESCRIPTION
**Pull request by tomwilkie**
Don't worry @brian-brazil, they're still snappy compressed.

Fixes #1982

### CODE REVIEW

File: storage/remote/client.go [hunk 1]
```diff
@@ -74,13 +65,39 @@ func (c *Client) Store(samples model.Samples) error {
...
+	if err != nil {
+		return err
+	}
+	defer httpResp.Body.Close()
+	if httpResp.StatusCode/100 != 2 {
+		return fmt.Errorf("server returned HTTP status %s", httpResp.Status)
+	}
+
+	buf.Reset()
+	if _, err = buf.ReadFrom(snappy.NewReader(httpResp.Body)); err != nil {
```
Comment (brian-brazil) [inline storage/remote/client.go]: Why are you reading the response?

**Comment by brian-brazil** [comment]
The example server needs updating too.

**Comment by juliusv** [comment]
Also unvendor all the gRPC dependencies drawn in by https://github.com/prometheus/prometheus/pull/1930...

**Comment by juliusv** [comment]
:+1: otherwise, if we decide to go this way.

File: storage/remote/client.go (hunk 1 above)
Comment (tomwilkie) [inline storage/remote/client.go]: It seemed like the right thing to do?  This is what grpc was doing.

I guess its meaning less so we could get rid of it.  I don't think there are any implication for forwards compatibility.

**Comment by brian-brazil** [comment]
:+1:

**Comment by juliusv** [comment]
Just learned that @fabxc is currently on the road, so let's give him a day to weigh in...

**Comment by juliusv** [comment]
@tomwilkie Could you rebase this? It has gotten 2 :+1:s and no objections in 2 days, so I would merge it soon...

**Comment by tomwilkie** [comment]
@juliusv there you go!

**Comment by fabxc** [comment]
If we cannot work on it, there's no use in gRPC.
Fine as it's still experimental anyway. We should track the upstream issue
and see whether we can go back to it before taking it into stable.

**Comment by tomwilkie** [comment]
Agreed!  I'd like to do a bit of work on gRPC to make the reverse proxying case work.

In the meantime, I'd love your input on #1957

**Comment by tamird** [comment]
FYI, it is possible to expose grpc services over HTTP using https://github.com/grpc-ecosystem/grpc-gateway. It is also possible to use protobuf-over-HTTP with grpc-gateway, though it requires a relatively small adapter: https://github.com/cockroachdb/cockroach/blob/master/util/proto_marshal.go

**Comment by achew22** [comment]
As one of the maintainers for grpc-ecosystem/grpc-gateway and a user of prometheus, I would be happy to help you if you have any questions on how stuff works.

**Comment by brian-brazil** [comment]
We're looking at using that for the read side of things, but hadn't considered it for the write side on performance side. How bad is the hit from using it and JSON?

**Comment by tamird** [comment]
Don't know, but like I said: you can get it to use protobuf-over-HTTP with a small adapter which you can fully crib from us: https://github.com/cockroachdb/cockroach/blob/master/util/proto_marshal.go.

**Comment by fabxc** [comment]
That's really interesting. I was not aware of that.

@juliusv @tomwilkie wdyt?

**Comment by brian-brazil** [comment]
I think we should hold off on making further changes until we're getting ready to make this non-experimental or have a concrete need for it. Changing around the format makes it harder to figure out the rest of the system.

**Pull request merged**